plugins/ti-dmc/fu-ti-dmc-device.c

```diff
@@ -0,0 +1,737 @@
+	g_autoptr(GPtrArray) chunks_payload = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 1, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 19, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 44, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 44, NULL);
+
+	/* get payload image */
+	fw_payload = fu_firmware_get_image_by_id_bytes(firmware, FU_FIRMWARE_ID_PAYLOAD, error);
+	if (fw_payload == NULL)
+		return FALSE;
+
+	/* SFWi */
+	if (!fu_ti_dmc_device_sfwi(self, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* write each SFWd block */
+	chunks_payload = fu_chunk_array_new_from_bytes(fw_payload,
+						       0x0, /* start_addr */
+						       0x0, /* page_sz */
+						       64);
+	if (!fu_ti_dmc_device_write_chunks(self,
+					   chunks_payload,
+					   fu_progress_get_child(progress),
+					   error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* SFWs with signature */
+	fw_sig = fu_firmware_get_image_by_id_bytes(firmware, FU_FIRMWARE_ID_SIGNATURE, error);
+	if (fw_sig == NULL)
+		return FALSE;
+	chunks_sig = fu_chunk_array_new_from_bytes(fw_sig,
+						   0x0, /* start_addr */
+						   0x0, /* page_sz */
+						   64);
+	if (!fu_ti_dmc_device_write_sfws_chunks(self,
+						chunks_sig,
+						fu_progress_get_child(progress),
+						error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* SFWs with pubkey */
+	fw_pubkey = fu_firmware_get_image_by_id_bytes(firmware, "pubkey", error);
+	if (fw_pubkey == NULL)
+		return FALSE;
+	chunks_pubkey = fu_chunk_array_new_from_bytes(fw_pubkey,
+						      0x0, /* start_addr */
+						      0x0, /* page_sz */
+						      64);
+	if (!fu_ti_dmc_device_write_sfws_chunks(self,
+						chunks_pubkey,
+						fu_progress_get_child(progress),
+						error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* success! */
+	return fu_ti_dmc_device_maybe_reboot(self, error);
+}
+
+static void
+fu_ti_dmc_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 57, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 43, "reload");
+}
+
+static void
+fu_ti_dmc_device_init(FuTiDmcDevice *self)
+{
+	fu_device_add_protocol(FU_DEVICE(self), "com.ti.dmc");
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_PLAIN);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_SIGNED_PAYLOAD);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_DUAL_IMAGE);
```

Comment:
Being dual image, I would think this is _probably_ also self recovery.  Is that the case?

Reply:
I'll ask, but I suspect so too.

Reply:
It is, fixed.